### ISSUE TITLE
Text inside <p> is overlaping floating image with max-width.

### ISSUE THREAD

**Issue by blacksvk** [servo/servo#11487]
![floatimg](https://cloud.githubusercontent.com/assets/10187350/15627207/11fe458e-24dd-11e6-9c3a-cada5dc5f3cd.png)

```
<!doctype html>
<html>
    <style>
        img{
            float: left;
            max-width:50%;
        }
    </style>
    <body>
        <div style="width: 600px;">
            <img src="http://upload.wikimedia.org/wikipedia/commons/thumb/0/0b/Skoda130Rally.jpg/512px-Skoda130Rally.jpg">
            <p style="background: lightgray;">
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ergo illi intellegunt quid Epicurus dicat, ego non intellego? Tum Lucius: Mihi vero ista valde probata sunt, quod item fratri puto. An eum discere ea mavis, quae cum plane perdidiceriti nihil sciat? Quid est enim aliud esse versutum? Et quod est munus, quod opus sapientiae? Quid de Platone aut de Democrito loquar?
            </p>
        </div>
    </body>
</html>
```

When you remove `max-width` -> everything's OK.

**Comment by chenpighead** [comment]
I'd like to look into this issue. I followed https://github.com/servo/servo/blob/master/docs/HACKING_QUICKSTART.md and was able to dumped debug log. Seems that setting max-width property to the floating img element: 1) won't set the flow to be dirty, 2) sets inline-size of the floating img element to be zero.

Either way may cause the float not being reflowed and readded. I'd like to do some further investigation, but not certain about next step. (since I'm very new to both rust and servo.) @jdm, could you give me some hints? Some hints like where might be the place that we would set the dirty flag for a floating flow? 
Or, if I'm not heading the wrong direction, please correct me as well.

**Comment by jdm** [comment]
I'm not sure I understand the bit about marking a flow as dirty. As far as I know (and based on [our documentation](https://github.com/servo/servo/wiki/Layout-Overview#incremental-style-calculation)), we only talk about marking nodes as dirty, not flows. Could you link the code you're talking about?

**Comment by jdm** [comment]
I will point out that [this code](https://dxr.mozilla.org/servo/rev/c4aaedab32eb906ff715b850b641d2a75536e73d/components/script/dom/htmlimageelement.rs#81) executes after the image data (either the metadata or the actual pixels) is received from the network, which marks the DOM node as dirty.

**Comment by chenpighead** [comment]
Element(HTMLElement(HTMLHtmlElement)): changed=false dirty=false dirty_descendants=false
  Element(HTMLElement(HTMLHeadElement)): changed=false dirty=false dirty_descendants=false
    Element(HTMLElement(HTMLStyleElement)): changed=false dirty=false dirty_descendants=false
      CharacterData(Text): changed=false dirty=false dirty_descendants=false
    CharacterData(Text): changed=false dirty=false dirty_descendants=false
  Element(HTMLElement(HTMLBodyElement)): changed=false dirty=false dirty_descendants=false
    CharacterData(Text): changed=false dirty=false dirty_descendants=false
    Element(HTMLElement(HTMLDivElement)): changed=false dirty=false dirty_descendants=false
      CharacterData(Text): changed=false dirty=false dirty_descendants=false
      Element(HTMLElement(HTMLImageElement)): changed=false dirty=false dirty_descendants=false
      CharacterData(Text): changed=false dirty=false dirty_descendants=false
      Element(HTMLElement(HTMLParagraphElement)): changed=false dirty=false dirty_descendants=false
        CharacterData(Text): changed=false dirty=false dirty_descendants=false
      CharacterData(Text): changed=false dirty=false dirty_descendants=false
    CharacterData(Text): changed=false dirty=false dirty_descendants=false
    CharacterData(Text): changed=false dirty=false dirty_descendants=false
    CharacterData(Text): changed=false dirty=false dirty_descendants=false
DEBUG:layout::block: assign_inline_sizes(block): assigning inline_size for flow
DEBUG:layout::block: Setting root position
DEBUG:layout::block: assign_inline_sizes(block): assigning inline_size for flow
DEBUG:layout::block: assign_inline_sizes(block): assigning inline_size for flow
DEBUG:layout::block: assign_inline_sizes(float): assigning inline_size for flow
DEBUG:layout::block: assign_inline_sizes_float -- inline_size: 0px
DEBUG:layout::block: assign_inline_sizes(block): assigning inline_size for flow
DEBUG:layout::inline: InlineFlow::assign_inline_sizes: floats in: [empty]
DEBUG:layout::inline: assign_block_size_inline: floats in: [empty]
DEBUG:layout::inline: LineBreaker: scanning for lines, 5 fragments
==
Above is part of the log I read before. So, I think you're right. Should be a node, not a flow. I misread the log.

I reread the log and found that seems we render text at first time, then trigger a reflow (or restyle? not sure.) The reflow/restyle did mark the img node as dirty. But while doing the text layout in layout::inline, computed inline size for float become empty. I suspect this is what causing the overlapping of the image and the text around it. Below is part of the log from the last reflow/restyle.

DEBUG:layout::construct: building flow for node: block none static_ Some(Element(Element))
DEBUG:layout::traversal: Constructed flow for 11b41c250: 117632010
DEBUG:layout::construct: building flow for node: block none static_ Some(Element(Element))
DEBUG:layout::traversal: Constructed flow for 11b41c1f0: 117632290
DEBUG:layout::construct: building flow for node: block none static_ Some(Element(Element))
DEBUG:layout::traversal: Constructed flow for 11b41c100: 117632510
DEBUG:style::traversal: [4] - 11544F000, and deleting BF.
DEBUG:layout::block: assign_inline_sizes(block): assigning inline_size for flow
DEBUG:layout::block: Setting root position
DEBUG:layout::block: assign_inline_sizes(block): assigning inline_size for flow
DEBUG:layout::block: assign_inline_sizes(block): assigning inline_size for flow
DEBUG:layout::block: assign_inline_sizes(float): assigning inline_size for flow
DEBUG:layout::block: assign_inline_sizes_float -- inline_size: 300px
DEBUG:layout::block: assign_inline_sizes(block): assigning inline_size for flow
DEBUG:layout::inline: InlineFlow::assign_inline_sizes: floats in: [empty]
DEBUG:layout::inline: assign_block_size_inline: floats in: [empty]
DEBUG:layout::inline: LineBreaker: scanning for lines, 5 fragments
==

For the correct case (try change max-width:50% to width:50%), I think the line
"DEBUG:layout::inline: InlineFlow::assign_inline_sizes: floats in: [empty]"
should be corrected to something like
"DEBUG:layout::inline: InlineFlow::assign_inline_sizes: floats in: offset=LogicalSize(H LTR, i-8px×b-162px) floats=max_block_start=Some(0px) floats=1 bounds=LogicalRect(H LTR, i300px×b206.25px, @ (i8px,b0px)) kind=Left"

I guess, the empty computed size of image node causes the text layout not to take the image node into consideration. I'll look into the code you mentioned. Thank you for the pointer.

**Comment by chenpighead** [comment]
Seems this bug can only be reproduced when setting percent value to max-width. If I set px/em to max-width, everything works fine. Interesting!!!!

**Comment by stshine** [comment]
I think this is fixed now by #14490.
